wavesrv/pkg/blockstore/blockstore.go

```diff
@@ -0,0 +1,449 @@
+	defer globalLock.Unlock()
+	for _, cacheEntry := range cache {
+		err := WriteFileToDB(ctx, *cacheEntry.Info)
+		if err != nil {
+			return err
+		}
+		for index, block := range cacheEntry.DataBlocks {
+			if block == nil || block.size == 0 {
+				continue
+			}
+			err := WriteDataBlockToDB(ctx, cacheEntry.Info.BlockId, cacheEntry.Info.Name, index, block.data)
+			if err != nil {
+				return err
+			}
+		}
+	}
+	cache = make(map[string]*CacheEntry)
+	return nil
+}
+
+// TODO, how does the cache handle race conditions with the read? If we are caching writes every second and the front end writes to the line, we would oveerrite it unless we read first
+// we would need a tcp like protocol if we need to do both reads and writes
+
+func ReadAt(ctx context.Context, blockId string, name string, p *[]byte, off int64) (int, error) {
+	bytesRead := 0
+	fInfo, err := Stat(ctx, blockId, name)
+	if err != nil {
+		return 0, fmt.Errorf("Read At err: %v", err)
+	}
+	if off > fInfo.Opts.MaxSize && fInfo.Opts.Circular {
+		numOver := off / fInfo.Opts.MaxSize
+		off = off - (numOver * fInfo.Opts.MaxSize)
+	}
+	if off > fInfo.Size {
+		return 0, fmt.Errorf("Read At error: tried to read past the end of the file")
+	}
+	endReadPos := math.Min(float64(int64(len(*p))+off), float64(fInfo.Size))
+	bytesToRead := int64(endReadPos) - off
+	curCacheNum := int(math.Floor(float64(off) / float64(MaxBlockSize)))
+	numCaches := int(math.Ceil(float64(bytesToRead) / float64(MaxBlockSize)))
+	cacheOffset := off - (int64(curCacheNum) * MaxBlockSize)
+	if (cacheOffset + int64(bytesToRead)) > MaxBlockSize {
+		numCaches += 1
+	}
+	log.Printf("mk1 %v %v", curCacheNum, numCaches)
+	for index := curCacheNum; index < curCacheNum+numCaches; index++ {
+		curCacheBlock, err := GetCacheBlock(ctx, blockId, name, index, true)
+		if err != nil {
+			return bytesRead, fmt.Errorf("Error getting cache block: %v", err)
+		}
+		cacheOffset := off - (int64(index) * MaxBlockSize)
+		bytesToReadFromCurCache := int(math.Min(float64(bytesToRead), float64(MaxBlockSize-cacheOffset)))
+		fileMaxSize := fInfo.Opts.MaxSize
+		maxReadSize := fileMaxSize - (int64(index) * MaxBlockSize)
+		b, err := ReadFromCacheBlock(ctx, blockId, name, curCacheBlock, p, int(cacheOffset), bytesToReadFromCurCache, bytesRead, maxReadSize)
+		log.Printf("reading; %v %v %v %v", b, off, bytesToReadFromCurCache, err)
+		bytesRead += b
+		bytesToRead -= int64(b)
+		off += int64(b)
+
+		if err != nil {
+			if err.Error() == MaxSizeError {
+				if fInfo.Opts.Circular {
+					off = 0
+					newP := (*p)[b:]
+					b, err := ReadAt(ctx, blockId, name, &newP, off)
+					bytesRead += b
+					log.Printf("mk max size read: %v %v %v", b, bytesRead, off)
+					if err != nil {
+						return bytesRead, err
+					}
+					break
+				}
+			} else {
+				return bytesRead, fmt.Errorf("Read from cache error: %v", err)
+			}
+		}
+	}
+	return bytesRead, nil
+}
+
+func AppendData(ctx context.Context, blockId string, name string, p []byte) (int, error) {
```

Comment:
I lock the entirety of AppendData, because append needs to get the offset for the end of the file and then update the end of the file without the end of the file changing